File: test/language/templates/cpp/StructTemplatedFieldTest.cpp
#include "gtest/gtest.h"
#include "templates/struct_templated_field/StructTemplatedField.h"
#include "zserio/RebindAlloc.h"

namespace templates
{
namespace struct_templated_field
{

using allocator_type = StructTemplatedField::allocator_type;
using string_type = zserio::string<allocator_type>;

TEST(StructTemplatedFieldTest, readWrite)
{
    StructTemplatedField structTemplatedField;
    structTemplatedField.setUint32Field(Field_uint32{42});
    structTemplatedField.setCompoundField(Field_Compound{Compound{42}});
    structTemplatedField.setStringField(Field_string{string_type{"string"}});

    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
    zserio::BitStreamWriter writer(bitBuffer);
    structTemplatedField.write(writer);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    StructTemplatedField readStructTemplatedField(reader);

    ASSERT_TRUE(structTemplatedField == readStructTemplatedField);
}

} // namespace struct_templated_field
} // namespace templates

File: test/language/templates/python/StructTemplatedFieldTest.py
import zserio

import Templates

class StructTemplatedFieldTest(Templates.TestCase):
    def testReadWrite(self):
        structTemplatedField = self.api.StructTemplatedField(
            self.api.Field_uint32(13),
            self.api.Field_Compound(self.api.Compound(42)),
            self.api.Field_string("string")
        )

        writer = zserio.BitStreamWriter()
        structTemplatedField.write(writer)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readStructTemplatedField = self.api.StructTemplatedField()
        readStructTemplatedField.read(reader)
        self.assertEqual(structTemplatedField, readStructTemplatedField)
